### PR TITLE
ci: test downstream

### PR DESCRIPTION
**Pull request by trentwatt**
## Description

adds `intg-test` that checks out saas and tests saas web

test pr against this branch:
- [#4616 upstream branch with shared changes](https://github.com/determined-ai/determined/pull/4616): should fail because it breaks saas tests [yep](https://app.circleci.com/pipelines/github/determined-ai/determined/24792/workflows/5f6feea2-8fb0-4c25-99f4-22a89b7b286c/jobs/805118) ✅
- [#4617 fork branch with shared changes](https://github.com/determined-ai/determined/pull/4617): should fail due to lack of CI permissions-- requires PR from upstream branch [yep](https://app.circleci.com/pipelines/github/determined-ai/determined/24791/workflows/d47eb68e-1e00-493b-a6a9-69fc7a091339/jobs/805081) ✅
- [#4618 fork branch w/o shared changes](https://github.com/determined-ai/determined/pull/4618): should pass because the downstream test is skipped [yep](https://app.circleci.com/pipelines/github/determined-ai/determined/24790/workflows/89a9dadb-dd38-4a9b-962b-9255384b218f/jobs/805064) ✅

for tests, i had to make a temporary switch to simulate it being `upstream/master`.  i will switch it back for merge, so it might trigger subsequent checks on the linked PRs with erroneous results. the `yeps` link to the actual workflow runs for future reference.

## Checklist

- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.
- [ ] If modifying `/webui/react/src/shared/` verify `make -C webui/react test-shared` passes.

### CODE REVIEW

File: webui/react/src/shared/bin/test.py
Comment (trentwatt) [inline webui/react/src/shared/bin/test.py:10]: don't know if this change was actually needed. will check.

Comment (hamidzr) [reply webui/react/src/shared/bin/test.py:10]: probably needed as it's a private repo. could we keep the spacing and quotations the same?

File: .circleci/config.yml
Comment (hamidzr) [inline .circleci/config.yml:1271]: hmm, would it fit better if it's categorized as an integration test? 

wdy about limiting this to run only on prs opened from upstream. if that's possible

Comment (trentwatt) [reply .circleci/config.yml:1271]: > hmm, would it fit better if it's categorized as an integration test?


yeah good call


> wdy about limiting this to run only on prs opened from upstream. if that's possible
i would rather have it run on forks as well, that way it "fails safe".  

if it automatically fails when it is not able to run, then it is easier to make sure that it is run, and we don't need any additional mechanisms (codeowners, checklist, PSAs)

File: webui/react/src/shared/bin/test.py
Comment (trentwatt) [reply webui/react/src/shared/bin/test.py:10]: leave `ssh` but change quotation marks?

or change to `https` and quotation marks and see if it works?

File: .circleci/config.yml
Comment (hamidzr) [inline .circleci/config.yml:1278]: use test-shared target?

File: .circleci/config.yml
Comment (hamidzr) [reply .circleci/config.yml:1271]: wouldn't that mean that there'd be a failing test for all prs originating from forks?

Comment (trentwatt) [reply .circleci/config.yml:1271]: can make it so where it skips for things that don't touch shared.

Comment (hamidzr) [reply .circleci/config.yml:1271]: perfect

Comment (trentwatt) [reply .circleci/config.yml]: ok it's up, will make some test PRs then put them in the description

File: webui/react/src/shared/bin/test.py
Comment (hamidzr) [reply webui/react/src/shared/bin/test.py]: if as is, it also works locally for people as well then either one sounds good.

File: .circleci/config.yml
Comment (hamidzr) [inline .circleci/config.yml:132]: nit: is this "must be run after.." needed ? 
maybe could be a comment 🤔 the name shows up in the ui

**Review by hamidzr — APPROVED** [review]
the changes lgtm/ thanks for setting this up. I asked for a review from @determined-ai/infrastructure just in case they notice something

File: .circleci/config.yml
Comment (trentwatt) [reply .circleci/config.yml]: hmm unfortunately it would be hard to test the skip because of how we get what's changed

`MERGE_BASE="$(git merge-base upstream/master HEAD)"`

emphasis on `upstream/master`

Comment (trentwatt) [reply .circleci/config.yml]: i will hard code it to `upstream/test-downstream` for testing purposes and switch back before merge.

File: .circleci/config.yml
Comment (trentwatt) [reply .circleci/config.yml]: ok i moved to comment.

the situation was they both do `git fetch upstream` and if you try to add an existing remote it errors. 

so just made it where you add the remote in one and then reuse it in the other. feels gross but w/e

Comment (trentwatt) [reply .circleci/config.yml]: could clean it up a bit, but maybe next time around lol.

**Review by loksonarius — APPROVED** [review]

File: .circleci/config.yml
Comment (trentwatt) [reply .circleci/config.yml]: it turns out putting it into `test-intg` made it no longer have access to remote.

@hamidzr do you think it's okay to do 
```
git remote add upstream https://github.com/determined-ai/determined 2>/dev/null
git fetch upstream
```

that way adding error for existing remote is supressed (but fetch fail will still cause error)

and we can get rid of dependency between `skip-webui` and `skip-docs`, making things simpler

**Pull request merged**
